napari/components/viewer_model.py

```diff
@@ -354,16 +354,57 @@ def _on_active_layer(self, event):
     def rounded_division(min_val, max_val, precision):
         return int(((min_val + max_val) / 2) / precision) * precision
 
-    def _on_layers_change(self):
+    def _on_layers_change(self, *, add_layer=False):
+        new_point = None
         if len(self.layers) == 0:
             self.dims.ndim = 2
             self.dims.reset()
         else:
             ranges = self.layers._ranges
+            # If the point of an axis is outside of the upper current range,
+            # set it to the layers range max value. dims.set_current_step
```

Comment:
It looks like `set_current_step` should also take care of clamping the upper limit: https://github.com/napari/napari/blob/e1f6779c2d117a34b0b3fcc8f5e3c169d0a20a8c/napari/components/dims.py#L303

But I tried replacing this block with `prev_point = self.dims.point` and found that all of the new tests failed, so I don't understand the explanation here.

Reply:
Ah, I think I see now. The following line prevents us from getting there: https://github.com/napari/napari/blob/e1f6779c2d117a34b0b3fcc8f5e3c169d0a20a8c/napari/components/dims.py#L298

If I remove that, then all your tests pass. It's possible that removing it causes other tests to fail, but my guess is probably not, as this looks like an optimization to avoid going through the root validation of `Dims` when `current_step` doesn't actually change.

I think we should either remove that line or maybe safer, do the equality check after clamping the passed values as with the scalar case. After that, I think we could remove your custom logic here to remove some code.

Reply:
I adjusted the set_current_step function by moving the check after the clamping :) 
So I removed the logic at this point.

Reply:
OK most of the logic, I needed to keep the n_layer == 1 logic on adding a layer to satisfy the tests in `napari/_tests/test_advanced.py` and `napari/layers/_tests/test_dask_layers.py`